### DISCUSSION TITLE
Problem with join, joinedload or contains_eager...

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#7981] [Usage Questions] [answered]
for query:
```python
def get_depm_join_user_by_id(self, id: int) -> Department:
    return self.db.query(Department).join(Department.users).options(
        contains_eager(Department.users)).filter(Department.id == id).all()
```
If i exclude a column,
how to use undefer() with `options(contains_eager(Department.users)`?
Or not undefer() but something else?
I tried:
```python
def get_depm_join_user_by_id(self, id: int) -> Department:
    return self.db.query(Department).join(Department.users).options(
        contains_eager(Department.users).undefer('password')).filter(Department.id == id).all()
```
or
```python
def get_depm_join_user_by_id(self, id: int) -> Department:
    return self.db.query(Department).join(Department.users).options(
        joinedload(Department.users).options(undefer('password'))).filter(Department.id == id).all()
```
or ......
But it didn't meet expectations
And i have a other question: how to use order_by() with `options(contains_eager(Department.users)` or `joinedload(Department.users)`?

**Comment by zzzeek** [comment, marked as answer]
Here's a complete example of contains_eager() + undefer() + orderby.   joinedload() cannot interact with order_by().

The objects at the end are detached, so the fact that the password prints successfully illustrates that the undefer works.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import contains_eager
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import deferred
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class Department(Base):
    __tablename__ = "department"

    id = Column(Integer, primary_key=True)
    users = relationship("User")


class User(Base):
    __tablename__ = "user"
    id = Column(Integer, primary_key=True)
    dept_id = Column(ForeignKey("department.id"))
    password = deferred(Column(String))


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add_all(
    [
        Department(users=[User(password="p1"), User(password="p2")]),
        Department(users=[User(password="p3"), User(password="p4")]),
    ]
)

s.commit()
s.close()

depts = (
    s.query(Department)
    .join(Department.users)
    .options(contains_eager(Department.users).undefer(User.password))
    .order_by(User.password.desc())
    .all()
)
s.close()
# all objects are detached
for d in depts:
    for user in d.users:
        print(user.password)
```

output:

``
SELECT user.password AS user_password, user.id AS user_id, user.dept_id AS user_dept_id, department.id AS department_id 
FROM department JOIN user ON department.id = user.dept_id ORDER BY user.password DESC
2022-05-01 22:01:17,395 INFO sqlalchemy.engine.Engine [generated in 0.00008s] ()
2022-05-01 22:01:17,395 INFO sqlalchemy.engine.Engine ROLLBACK
p4
p3
p2
p1
```

**Comment by wu-clan** [comment]


**Reply by CaselIT** [reply]
> undefer(): Uncertain properties, Exclude or include? Do not load by default? But loading it will not throw an exception

`undefer`, like the name implies (or at least tries to), does the opposite of `defer`, so it includes a column in the result. This is used for columns that are marked `deferred()` in the model definition

**Reply by wu-clan** [reply]
@CaselIT I have a deeper understanding of them. At present, my question is that `under` Is it an extension of `load_only`?

**Reply by CaselIT** [reply]
I would not say an extension, they do two different things.
Undefer controls the behavior of column that are normally deferred, while load_only allow specifying what columns to load independently of the fact that they are deferred or not.

**Reply by wu-clan** [reply]
Thank you very much for your answer, which makes me very clear
